lib/ads-adal-library/src/engine/AzureAuth.ts

```diff
@@ -126,7 +126,7 @@ export abstract class AzureAuth {
 			const remainingTime = expiry - currentTime;
 			const maxTolerance = 2 * 60; // two minutes
 
-			if (remainingTime < maxTolerance) {
+			if ((remainingTime < maxTolerance) && (remainingTime > 0)) {
```

Comment:
Why only when remainingTime > 0? If the expiration time has passed we need to refresh as well.

Reply:
1. Yes, actually it is fixing #17409 and #17375
2. The issue is that sometime the refreshToken is outdated as well (e.g. Password change  #17409 or VSCode finds some old cached token #17375). In these cases it cannot refresh the token and returns an invalid token.

Reply:
> If the expiration time has passed we need to refresh as well right? Yes, the issue is that we might try to refresh with an invalid token, due to it only being valid for 24h hours. Or due to a password change the refresh token is invalid, or it has been revoked. In these cases we end up without a valid token and return undefined;

Reply:
Refreshing the token can fail for a number of reasons - including some like token revocation or password change that can happen at any time. So what seems like a better overall fix is to always try to refresh the token but then be better about handling errors (which it seems like we're not doing right now). Which would include ensuring that the old token is cleared out if it's invalid. 

@cssuh @kburtram Are these scenarios something you've had a chance to look into?